### ISSUE TITLE
Changes in version 1.1.5 make glmmTMB no longer return SE for specific models

### ISSUE THREAD

**Issue by strengejacke** [glmmTMB/glmmTMB#880]
I'm preparing an update and while running tests, I realized that some behaviour of _glmmTMB_ has changed. See following example, where SE etc. could be computed for version 1.1.4, but no longer for 1.1.5.

## glmmTMB 1.1.4

``` r
library(glmmTMB)
library(lme4)
#> Loading required package: Matrix

data(sleepstudy)
set.seed(123)
sleepstudy$Months <- sample(1:4, nrow(sleepstudy), TRUE)

m <- glmmTMB(Reaction ~ Days + (0 + Days + Months | Subject), data = sleepstudy)
#> Warning in fitTMB(TMBStruc): Model convergence problem; non-positive-definite
#> Hessian matrix. See vignette('troubleshooting')
summary(m)
#>  Family: gaussian  ( identity )
#> Formula:          Reaction ~ Days + (0 + Days + Months | Subject)
#> Data: sleepstudy
#> 
#>      AIC      BIC   logLik deviance df.resid 
#>       NA       NA       NA       NA      174 
#> 
#> Random effects:
#> 
#> Conditional model:
#>  Groups   Name   Variance  Std.Dev.  Corr 
#>  Subject  Days   1.479e-09 3.845e-05      
#>           Months 1.659e+02 1.288e+01 1.00 
#>  Residual        1.170e+03 3.420e+01      
#> Number of obs: 180, groups:  Subject, 18
#> 
#> Dispersion estimate for gaussian family (sigma^2): 1.17e+03 
#> 
#> Conditional model:
#>             Estimate Std. Error z value Pr(>|z|)    
#> (Intercept) 250.3959     6.8159   36.74   <2e-16 ***
#> Days         10.7952     0.8948   12.06   <2e-16 ***
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

<sup>Created on 2022-11-19 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

## glmmTMB 1.1.5

``` r
library(glmmTMB)
library(lme4)
#> Loading required package: Matrix

data(sleepstudy)
set.seed(123)
sleepstudy$Months <- sample(1:4, nrow(sleepstudy), TRUE)

m <- glmmTMB(Reaction ~ Days + (0 + Days + Months | Subject), data = sleepstudy)
#> Warning in fitTMB(TMBStruc): Model convergence problem; non-positive-definite
#> Hessian matrix. See vignette('troubleshooting')
summary(m)
#>  Family: gaussian  ( identity )
#> Formula:          Reaction ~ Days + (0 + Days + Months | Subject)
#> Data: sleepstudy
#> 
#>      AIC      BIC   logLik deviance df.resid 
#>       NA       NA       NA       NA      174 
#> 
#> Random effects:
#> 
#> Conditional model:
#>  Groups   Name   Variance  Std.Dev.  Corr 
#>  Subject  Days   1.479e-09 3.845e-05      
#>           Months 1.659e+02 1.288e+01 1.00 
#>  Residual        1.170e+03 3.420e+01      
#> Number of obs: 180, groups:  Subject, 18
#> 
#> Dispersion estimate for gaussian family (sigma^2): 1.17e+03 
#> 
#> Conditional model:
#>             Estimate Std. Error z value Pr(>|z|)
#> (Intercept)    250.4        NaN     NaN      NaN
#> Days            10.8        NaN     NaN      NaN
```

<sup>Created on 2022-11-19 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by bbolker** [comment]
I can't replicate this (see `sessionInfo()` output below); I assume that it's a matter of an unstable fit that is slipping between "Hessian reported as non-pos-def but can still get Wald estimates" and "can't get Wald estimates".  Will try some experiments with different seeds later on ...

```
R Under development (unstable) (2022-11-02 r83236)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Pop!_OS 22.04 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=en_CA.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_CA.UTF-8        LC_COLLATE=en_CA.UTF-8    
 [5] LC_MONETARY=en_CA.UTF-8    LC_MESSAGES=en_CA.UTF-8   
 [7] LC_PAPER=en_CA.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_CA.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] lme4_1.1-31   Matrix_1.5-3  glmmTMB_1.1.5

loaded via a namespace (and not attached):
 [1] codetools_0.2-18    numDeriv_2016.8-1.1 multcomp_1.4-20    
 [4] lattice_0.20-45     TH.data_1.1-1       coda_0.19-4        
 [7] splines_4.3.0       estimability_1.4.1  zoo_1.8-11         
[10] emmeans_1.8.2       mvtnorm_1.1-3       TMB_1.9.1          
[13] xtable_1.8-4        nloptr_2.0.3        sandwich_3.0-2     
[16] grid_4.3.0          compiler_4.3.0      boot_1.3-28        
[19] tools_4.3.0         nlme_3.1-160        minqa_1.2.5        
[22] survival_3.4-0      Rcpp_1.0.9          MASS_7.3-58.1      
```

**Comment by strengejacke** [comment]
I have included session info for both reprex's now.

``` r
library(glmmTMB)

data(sleepstudy, package = "lme4")
set.seed(123)
sleepstudy$Months <- sample(1:4, nrow(sleepstudy), TRUE)

m <- glmmTMB(Reaction ~ Days + (0 + Days + Months | Subject), data = sleepstudy)
#> Warning in fitTMB(TMBStruc): Model convergence problem; non-positive-definite
#> Hessian matrix. See vignette('troubleshooting')
summary(m)
#>  Family: gaussian  ( identity )
#> Formula:          Reaction ~ Days + (0 + Days + Months | Subject)
#> Data: sleepstudy
#> 
#>      AIC      BIC   logLik deviance df.resid 
#>       NA       NA       NA       NA      174 
#> 
#> Random effects:
#> 
#> Conditional model:
#>  Groups   Name   Variance  Std.Dev.  Corr 
#>  Subject  Days   1.479e-09 3.845e-05      
#>           Months 1.659e+02 1.288e+01 1.00 
#>  Residual        1.170e+03 3.420e+01      
#> Number of obs: 180, groups:  Subject, 18
#> 
#> Dispersion estimate for gaussian family (sigma^2): 1.17e+03 
#> 
#> Conditional model:
#>             Estimate Std. Error z value Pr(>|z|)    
#> (Intercept) 250.3959     6.8159   36.74   <2e-16 ***
#> Days         10.7952     0.8948   12.06   <2e-16 ***
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.2 (2022-10-31 ucrt)
#>  os       Windows 10 x64 (build 19045)
#>  system   x86_64, mingw32
#>  ui       RTerm
#>  language (EN)
#>  collate  German_Germany.utf8
#>  ctype    German_Germany.utf8
#>  tz       Europe/Berlin
#>  date     2022-11-19
#>  pandoc   2.19.2 @ C:/PROGRA~1/Pandoc/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package      * version    date (UTC) lib source
#>  boot           1.3-28     2021-05-03 [2] CRAN (R 4.2.2)
#>  cli            3.4.1      2022-09-23 [1] CRAN (R 4.2.1)
#>  coda           0.19-4     2020-09-30 [1] CRAN (R 4.2.0)
#>  codetools      0.2-18     2020-11-04 [2] CRAN (R 4.2.2)
#>  digest         0.6.30     2022-10-18 [1] CRAN (R 4.2.1)
#>  emmeans        1.8.2      2022-10-27 [1] CRAN (R 4.2.1)
#>  estimability   1.4.1      2022-08-05 [1] CRAN (R 4.2.1)
#>  evaluate       0.18       2022-11-07 [1] CRAN (R 4.2.2)
#>  fastmap        1.1.0      2021-01-25 [1] CRAN (R 4.2.0)
#>  fs             1.5.2      2021-12-08 [1] CRAN (R 4.2.0)
#>  glmmTMB      * 1.1.4      2022-07-12 [1] CRAN (R 4.2.2)
#>  glue           1.6.2      2022-02-24 [1] CRAN (R 4.2.0)
#>  highr          0.9        2021-04-16 [1] CRAN (R 4.2.0)
#>  htmltools      0.5.3      2022-07-18 [1] CRAN (R 4.2.1)
#>  knitr          1.41       2022-11-18 [1] CRAN (R 4.2.2)
#>  lattice        0.20-45    2021-09-22 [1] CRAN (R 4.2.0)
#>  lifecycle      1.0.3      2022-10-07 [1] CRAN (R 4.2.1)
#>  lme4           1.1-31     2022-11-01 [1] CRAN (R 4.2.2)
#>  magrittr       2.0.3      2022-03-30 [1] CRAN (R 4.2.0)
#>  MASS           7.3-58.1   2022-08-03 [2] CRAN (R 4.2.2)
#>  Matrix         1.5-3      2022-11-11 [1] CRAN (R 4.2.2)
#>  minqa          1.2.5      2022-10-19 [1] CRAN (R 4.2.1)
#>  multcomp       1.4-20     2022-08-07 [1] CRAN (R 4.2.1)
#>  mvtnorm        1.1-3      2021-10-08 [1] CRAN (R 4.2.0)
#>  nlme           3.1-160    2022-10-10 [2] CRAN (R 4.2.2)
#>  nloptr         2.0.3      2022-05-26 [1] CRAN (R 4.2.0)
#>  numDeriv       2016.8-1.1 2019-06-06 [1] CRAN (R 4.2.0)
#>  purrr          0.3.5      2022-10-06 [1] CRAN (R 4.2.1)
#>  R.cache        0.16.0     2022-07-21 [1] CRAN (R 4.2.1)
#>  R.methodsS3    1.8.2      2022-06-13 [1] CRAN (R 4.2.0)
#>  R.oo           1.25.0     2022-06-12 [1] CRAN (R 4.2.0)
#>  R.utils        2.12.2     2022-11-11 [1] CRAN (R 4.2.2)
#>  Rcpp           1.0.9      2022-07-08 [1] CRAN (R 4.2.1)
#>  reprex         2.0.2      2022-08-17 [1] CRAN (R 4.2.1)
#>  rlang          1.0.6      2022-09-24 [1] CRAN (R 4.2.1)
#>  rmarkdown      2.18       2022-11-09 [1] CRAN (R 4.2.2)
#>  sandwich       3.0-2      2022-06-15 [1] CRAN (R 4.2.0)
#>  sessioninfo    1.2.2      2021-12-06 [1] CRAN (R 4.2.0)
#>  stringi        1.7.8      2022-07-11 [1] CRAN (R 4.2.1)
#>  stringr        1.4.1      2022-08-20 [1] CRAN (R 4.2.1)
#>  styler         1.8.1      2022-11-07 [1] CRAN (R 4.2.2)
#>  survival       3.4-0      2022-08-09 [1] CRAN (R 4.2.1)
#>  TH.data        1.1-1      2022-04-26 [1] CRAN (R 4.2.0)
#>  TMB            1.9.1      2022-08-16 [1] CRAN (R 4.2.2)
#>  withr          2.5.0      2022-03-03 [1] CRAN (R 4.2.0)
#>  xfun           0.35       2022-11-16 [1] CRAN (R 4.2.2)
#>  xtable         1.8-4      2019-04-21 [1] CRAN (R 4.2.0)
#>  yaml           2.3.6      2022-10-18 [1] CRAN (R 4.2.1)
#>  zoo            1.8-11     2022-09-17 [1] CRAN (R 4.2.1)
#> 
#>  [1] C:/Users/mail/AppData/Local/R/win-library/4.2
#>  [2] C:/Program Files/R/R-4.2.2/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```


``` r
library(glmmTMB)

data(sleepstudy, package = "lme4")
set.seed(123)
sleepstudy$Months <- sample(1:4, nrow(sleepstudy), TRUE)

m <- glmmTMB(Reaction ~ Days + (0 + Days + Months | Subject), data = sleepstudy)
#> Warning in fitTMB(TMBStruc): Model convergence problem; non-positive-definite
#> Hessian matrix. See vignette('troubleshooting')
summary(m)
#>  Family: gaussian  ( identity )
#> Formula:          Reaction ~ Days + (0 + Days + Months | Subject)
#> Data: sleepstudy
#> 
#>      AIC      BIC   logLik deviance df.resid 
#>       NA       NA       NA       NA      174 
#> 
#> Random effects:
#> 
#> Conditional model:
#>  Groups   Name   Variance  Std.Dev.  Corr 
#>  Subject  Days   1.479e-09 3.845e-05      
#>           Months 1.659e+02 1.288e+01 1.00 
#>  Residual        1.170e+03 3.420e+01      
#> Number of obs: 180, groups:  Subject, 18
#> 
#> Dispersion estimate for gaussian family (sigma^2): 1.17e+03 
#> 
#> Conditional model:
#>             Estimate Std. Error z value Pr(>|z|)
#> (Intercept)    250.4        NaN     NaN      NaN
#> Days            10.8        NaN     NaN      NaN

sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.2 (2022-10-31 ucrt)
#>  os       Windows 10 x64 (build 19045)
#>  system   x86_64, mingw32
#>  ui       RTerm
#>  language (EN)
#>  collate  German_Germany.utf8
#>  ctype    German_Germany.utf8
#>  tz       Europe/Berlin
#>  date     2022-11-19
#>  pandoc   2.19.2 @ C:/PROGRA~1/Pandoc/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package      * version    date (UTC) lib source
#>  boot           1.3-28     2021-05-03 [2] CRAN (R 4.2.2)
#>  cli            3.4.1      2022-09-23 [1] CRAN (R 4.2.1)
#>  coda           0.19-4     2020-09-30 [1] CRAN (R 4.2.0)
#>  codetools      0.2-18     2020-11-04 [2] CRAN (R 4.2.2)
#>  digest         0.6.30     2022-10-18 [1] CRAN (R 4.2.1)
#>  emmeans        1.8.2      2022-10-27 [1] CRAN (R 4.2.1)
#>  estimability   1.4.1      2022-08-05 [1] CRAN (R 4.2.1)
#>  evaluate       0.18       2022-11-07 [1] CRAN (R 4.2.2)
#>  fastmap        1.1.0      2021-01-25 [1] CRAN (R 4.2.0)
#>  fs             1.5.2      2021-12-08 [1] CRAN (R 4.2.0)
#>  glmmTMB      * 1.1.5      2022-11-16 [1] CRAN (R 4.2.2)
#>  glue           1.6.2      2022-02-24 [1] CRAN (R 4.2.0)
#>  highr          0.9        2021-04-16 [1] CRAN (R 4.2.0)
#>  htmltools      0.5.3      2022-07-18 [1] CRAN (R 4.2.1)
#>  knitr          1.41       2022-11-18 [1] CRAN (R 4.2.2)
#>  lattice        0.20-45    2021-09-22 [1] CRAN (R 4.2.0)
#>  lifecycle      1.0.3      2022-10-07 [1] CRAN (R 4.2.1)
#>  lme4           1.1-31     2022-11-01 [1] CRAN (R 4.2.2)
#>  magrittr       2.0.3      2022-03-30 [1] CRAN (R 4.2.0)
#>  MASS           7.3-58.1   2022-08-03 [2] CRAN (R 4.2.2)
#>  Matrix         1.5-3      2022-11-11 [1] CRAN (R 4.2.2)
#>  minqa          1.2.5      2022-10-19 [1] CRAN (R 4.2.1)
#>  multcomp       1.4-20     2022-08-07 [1] CRAN (R 4.2.1)
#>  mvtnorm        1.1-3      2021-10-08 [1] CRAN (R 4.2.0)
#>  nlme           3.1-160    2022-10-10 [2] CRAN (R 4.2.2)
#>  nloptr         2.0.3      2022-05-26 [1] CRAN (R 4.2.0)
#>  numDeriv       2016.8-1.1 2019-06-06 [1] CRAN (R 4.2.0)
#>  purrr          0.3.5      2022-10-06 [1] CRAN (R 4.2.1)
#>  R.cache        0.16.0     2022-07-21 [1] CRAN (R 4.2.1)
#>  R.methodsS3    1.8.2      2022-06-13 [1] CRAN (R 4.2.0)
#>  R.oo           1.25.0     2022-06-12 [1] CRAN (R 4.2.0)
#>  R.utils        2.12.2     2022-11-11 [1] CRAN (R 4.2.2)
#>  Rcpp           1.0.9      2022-07-08 [1] CRAN (R 4.2.1)
#>  reprex         2.0.2      2022-08-17 [1] CRAN (R 4.2.1)
#>  rlang          1.0.6      2022-09-24 [1] CRAN (R 4.2.1)
#>  rmarkdown      2.18       2022-11-09 [1] CRAN (R 4.2.2)
#>  sandwich       3.0-2      2022-06-15 [1] CRAN (R 4.2.0)
#>  sessioninfo    1.2.2      2021-12-06 [1] CRAN (R 4.2.0)
#>  stringi        1.7.8      2022-07-11 [1] CRAN (R 4.2.1)
#>  stringr        1.4.1      2022-08-20 [1] CRAN (R 4.2.1)
#>  styler         1.8.1      2022-11-07 [1] CRAN (R 4.2.2)
#>  survival       3.4-0      2022-08-09 [1] CRAN (R 4.2.1)
#>  TH.data        1.1-1      2022-04-26 [1] CRAN (R 4.2.0)
#>  TMB            1.9.1      2022-08-16 [1] CRAN (R 4.2.2)
#>  withr          2.5.0      2022-03-03 [1] CRAN (R 4.2.0)
#>  xfun           0.35       2022-11-16 [1] CRAN (R 4.2.2)
#>  xtable         1.8-4      2019-04-21 [1] CRAN (R 4.2.0)
#>  yaml           2.3.6      2022-10-18 [1] CRAN (R 4.2.1)
#>  zoo            1.8-11     2022-09-17 [1] CRAN (R 4.2.1)
#> 
#>  [1] C:/Users/mail/AppData/Local/R/win-library/4.2
#>  [2] C:/Program Files/R/R-4.2.2/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

<sup>Created on 2022-11-19 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by bbolker** [comment]
Test code:

```r
library(glmmTMB)
library(tidyverse)
library(broom.mixed)

data(sleepstudy, package = "lme4")

res <- list()
nsim <- 500
pb <- txtProgressBar(max = nsim, style = 3)
for (i in 1:nsim) {
    setTxtProgressBar(pb, i)
    set.seed(1000+i)
    sleepstudy$Months <- sample(1:4, nrow(sleepstudy), TRUE)
    res[[i]] <- glmmTMB(Reaction ~ Days + (0 + Days + Months | Subject), data = sleepstudy)
}
close(pb)
get_stderr <- function(x) { tidy(x, effects = "fixed") |> select(term, std.error) |>
                               pivot_wider(names_from = "term", values_from = "std.error") }
res <- purrr::map_dfr(res, get_stderr)
sum(sapply(res, function(x) !x$sdr$pdHess))   ## 493
colSums(is.na(as.matrix(res)))   ## 25
```

When I run this code on Pop!OS Linux with glmmTMB 1.1.5 I get 493/500 non-PD Hessians (based on glmmTMB's internal test, which I think uses `optimHess`) and 25/500 cases with NaN standard errors.  With glmmTMB 1.1.4 I get 493/500 and 32/500 respectively.

I could go through and `git bisect` for some of the cases that switch between releases if we wanted to know exactly what changes were responsible ...

**Comment by strengejacke** [comment]
Ok, so you would say this issue is not indicating any new bug, but something that can happen on random occurrence? If this is no real bug, we can close this issue.

**Comment by bbolker** [comment]
By definition any user-visible change between releases should be noted (and at least understood) ... I'm pretty sure I wouldn't classify this as a bug, but a "user-visible change". But would still sort of like to know what's going on ...

**Comment by bbolker** [comment]
Figured it out, after some `git bisect` effort. We switched to a new autodiff library in this commit a9391c59f7, merged here ef85ab8e217: I know `TMBad` is newer and has some additional capabilities beyond `CppAD` (the original AD library used), but you'd have to ask @kaskr for details about the differences (lots of new development/optimizations appear to be `TMBad`-specific).  It's easy to revert to the old behaviour (at least for one particular seed I tried!) by commenting out the following line in `src/Makevars`:

```
PKG_CPPFLAGS = -DTMBAD_FRAMEWORK
```

This definitely deserves a (retroactive) note in the NEWS file!
